server/s3_put_multiobject_copy_action.cc

```diff
@@ -0,0 +1,912 @@
+  part_metadata->set_pvid(motr_writer->get_ppvid());
+  new_oid_str = S3M0Uint128Helper::to_string(new_object_oid);
+
+  add_object_oid_to_probable_dead_oid_list();
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3PutMultipartCopyAction::create_part_object_failed() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  if (check_shutdown_and_rollback()) {
+    s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+    return;
+  }
+  s3_timer.stop();
+  const auto mss = s3_timer.elapsed_time_in_millisec();
+  LOG_PERF("create_part_object_failed_ms", request_id.c_str(), mss);
+  s3_stats_timing("create_part_object_failed", mss);
+
+  s3_copy_part_action_state = S3CopyPartActionState::newObjOidCreationFailed;
+
+  if (motr_writer->get_state() == S3MotrWiterOpState::failed_to_launch) {
+    s3_log(S3_LOG_ERROR, request_id, "Create object failed.\n");
+    set_s3_error("ServiceUnavailable");
+  } else {
+    s3_log(S3_LOG_WARN, request_id, "Create object failed.\n");
+    // Any other error report failure.
+    set_s3_error("InternalError");
+  }
+  send_response_to_s3_client();
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+// Initiate copy of source part object to the destination
+void S3PutMultipartCopyAction::initiate_part_copy() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  max_parallel_copy = MAX_PARALLEL_COPY;
+  if (additional_object_metadata->get_number_of_fragments() == 0) {
+    copy_part_object();
+  } else {
+    s3_copy_part_action_state = S3CopyPartActionState::validationFailed;
+    set_s3_error("NotImplemented");
+    send_response_to_s3_client();
+  }
+}
+
+// Copy source part object to destination
+void S3PutMultipartCopyAction::copy_part_object() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+
+  if (!total_data_to_stream) {
+    s3_log(S3_LOG_DEBUG, stripped_request_id, "Source object is empty");
+    next();
+    return;
+  }
+  bool f_success = false;
+  try {
+    object_data_copier.reset(new S3ObjectDataCopier(
+        request, motr_writer, motr_reader_factory, s3_motr_api));
+
+    object_data_copier->copy(
+        additional_object_metadata->get_oid(), total_data_to_stream,
+        additional_object_metadata->get_layout_id(),
+        additional_object_metadata->get_pvid(),
+        std::bind(&S3PutMultipartCopyAction::copy_part_object_cb, this),
+        std::bind(&S3PutMultipartCopyAction::copy_part_object_success, this),
+        std::bind(&S3PutMultipartCopyAction::copy_part_object_failed, this));
+    f_success = true;
+  }
+  catch (const std::exception& ex) {
+    s3_log(S3_LOG_ERROR, stripped_request_id, "%s", ex.what());
+  }
+  catch (...) {
+    s3_log(S3_LOG_ERROR, stripped_request_id, "Non-standard C++ exception");
+  }
+  if (!f_success) {
+    object_data_copier.reset();
+
+    set_s3_error("InternalError");
+    send_response_to_s3_client();
+    return;
+  }
+  if (total_data_to_stream > MINIMUM_ALLOWED_PART_SIZE) {
```

Comment:
If this is the last part that we are calling UploadPartCopy API, this check  (> 5MB) is not required. Actually, we don't know whether this is the last part (we would know this only during PostComplete API, and not now). 
User can use UploadPartCopy API to create a part with size < 5MB, which is allowed if this the last part. But as we don't know here, we should not have any lower size-check restriction here.

Reply:
@DPG17 You are right, I have removed this check.